### PR TITLE
[DROOLS-6618] Create single DTO to propagate compilation information

### PR DESCRIPTION
**Pull request by gitgabrio**
@danielezonca @kostola  @jiripetrlik

See https://issues.redhat.com/browse/DROOLS-6618

### CODE REVIEW

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/Constants.java [hunk 1]
```diff
@@ -40,8 +40,11 @@
     public static final String MISSING_METHOD_REFERENCE_TEMPLATE = "Missing method reference '%s' in %s";
     public static final String PACKAGE_CLASS_TEMPLATE = "%s.%s";
     public static final String WRONG_EXPRESSION_TEMPLATE = "Retrieved %s (%s) while a %s was expected from \n%s";
+    public static final String EXPECTING_HAS_KNOWLEDGEBUILDER_TEMPLATE = "Expecting HasKnowledgeBuilder, received %s";
```
Comment (jiripetrlik) [inline kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/Constants.java:43]: Please format properly.

**Review by jiripetrlik — APPROVED** [review]

File: kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/Constants.java (hunk 1 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-commons/src/main/java/org/kie/pmml/commons/Constants.java:43]: @jiripetrlik 
Done (I guess)

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-api/src/main/java/org/kie/pmml/compiler/api/utils/ModelUtils.java [hunk 2]
```diff
@@ -297,7 +297,7 @@ public static DATA_TYPE getDATA_TYPE(final List<Field<?>> fields, String targetF
      */
     public static List<org.kie.pmml.api.models.MiningField> convertToKieMiningFieldList(final MiningSchema toConvert,
                                                                                         final List<Field<?>> fields) {
-        if (toConvert == null) {
+        if (toConvert == null || !toConvert.hasMiningFields()) {
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-api/src/main/java/org/kie/pmml/compiler/api/utils/ModelUtils.java:300]: Is this a bugfix?

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/utils/ModelUtilsTest.java [hunk 3]
```diff
@@ -55,6 +55,7 @@
 import org.kie.pmml.commons.model.KiePMMLTarget;
 import org.kie.pmml.commons.model.KiePMMLTargetValue;
 import org.kie.pmml.commons.model.tuples.KiePMMLNameOpType;
+import org.kie.pmml.compiler.api.utils.ModelUtils;
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/utils/ModelUtilsTest.java]: Partial refactoring, you moved the original class but not the test (still in commons package)

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-api/src/main/java/org/kie/pmml/compiler/api/dto/CommonCompilationDTO.java [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+            targetDataField = null;
+        }
+        opType = targetDataField != null ? targetDataField.getOpType() : null;
+    }
+
+    public CommonCompilationDTO(final String packageName,
+                                final PMML pmml,
+                                final T model,
+                                final HasClassLoader hasClassloader) {
+        this(pmml, model, hasClassloader, getSanitizedPackageName(String.format(PACKAGE_CLASS_TEMPLATE, packageName,
```
Comment (danielezonca) [inline kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-api/src/main/java/org/kie/pmml/compiler/api/dto/CommonCompilationDTO.java:133]: Why `getSanitizedPackageName` is only here and not the base constructor?

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-api/src/main/java/org/kie/pmml/compiler/api/utils/ModelUtils.java (hunk 2 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-api/src/main/java/org/kie/pmml/compiler/api/utils/ModelUtils.java:300]: @danielezonca 
It is a further check. Those objects (DataDictionary, MiningSChema, Output, etc.) have an "has...." method to verify if the contained collection is null or not.

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-api/src/main/java/org/kie/pmml/compiler/api/dto/CommonCompilationDTO.java (hunk 4 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-api/src/main/java/org/kie/pmml/compiler/api/dto/CommonCompilationDTO.java:133]: @danielezonca 
The protected constructors are meant to receive already-formatted, not-to-be-changed, package name. See Javadoc

Comment (danielezonca) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-api/src/main/java/org/kie/pmml/compiler/api/dto/CommonCompilationDTO.java:133]: Ok so probably the [other constructor](https://github.com/kiegroup/drools/pull/3896/files/d835542e3b7f0667dac20c2031705ea341732911#diff-b29409c1f8f21ed23b42ba88a49cd0d798557e5d1e1a247fd62004ff787336ebR94) is still public by mistake 🤔

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/utils/ModelUtilsTest.java (hunk 3 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-commons/src/test/java/org/kie/pmml/compiler/commons/utils/ModelUtilsTest.java]: @danielezonca 
done

File: kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-api/src/main/java/org/kie/pmml/compiler/api/dto/CommonCompilationDTO.java (hunk 4 above)
Comment (gitgabrio) [reply kie-pmml-trusty/kie-pmml-compiler/kie-pmml-compiler-api/src/main/java/org/kie/pmml/compiler/api/dto/CommonCompilationDTO.java:133]: It is used inside `AbstractSpecificCompilationDTO` for same reason

**Comment by gitgabrio** [comment]
Jenkins retest this

**Review by danielezonca — APPROVED** [review]

**Comment by gitgabrio** [comment]
Jenkins run cdb

**Pull request merged**